neo4j/session_with_context.go

```diff
@@ -545,49 +564,105 @@ func (s *sessionWithContext) getServers(mode idb.AccessMode) func() []string {
+		}
+	}
+
+	conn.SetPinHomeDatabaseCallback(s.pinHomeDatabaseCallback)
+
 	// Select database on server
 	if s.config.DatabaseName != idb.DefaultDatabase {
-		dbSelector, ok := conn.(idb.DatabaseSelector)
-		if !ok {
-			s.pool.Return(ctx, conn)
-			return nil, &UsageError{Message: "Database does not support multi-database"}
+		if err := s.selectDatabase(ctx, conn); err != nil {
+			return nil, err
 		}
-		dbSelector.SelectDatabase(s.config.DatabaseName)
 	}
 
 	return conn, nil
 }
 
+// applyConnectionTimeout sets a timeout on the context if configured.
+func (s *sessionWithContext) applyConnectionTimeout(ctx context.Context) (context.Context, context.CancelFunc) {
+	timeout := s.driverConfig.ConnectionAcquisitionTimeout
+	if timeout > 0 {
+		ctx, cancel := context.WithTimeout(ctx, timeout)
+		deadline, _ := ctx.Deadline()
+		s.log.Debugf(log.Session, s.logId, "connection acquisition timeout is %s, resolved deadline is: %s", timeout, deadline)
+		return ctx, cancel
+	}
+	if deadline, ok := ctx.Deadline(); ok {
+		s.log.Debugf(log.Session, s.logId, "connection acquisition user-provided deadline is: %s", deadline)
+	}
+	return ctx, nil
+}
+
+// getServerList resolves the server list based on the session configuration.
+// It returns a list of servers, a boolean indicating whether the home database guess was used, and an error if resolution fails.
+func (s *sessionWithContext) getServerList(ctx context.Context, mode idb.AccessMode) ([]string, bool, error) {
+	if s.config.DatabaseName != "" {
+		if err := s.resolveHomeDatabase(ctx); err != nil {
+			return nil, false, err
+		}
+		serverList, err := s.getOrUpdateServers(ctx, mode, false)
+		return serverList, false, err
+	}
+
+	// Use cached home database if available
+	if s.homeDbGuess != "" && s.router.GetTable(s.homeDbGuess) != nil && s.cache.IsEnabled() {
+		serverList, err := s.getOrUpdateServers(ctx, mode, true)
+		return serverList, true, err
+	}
+
+	// Resolve home database if cache is unavailable, there's no guess, or a table doesn't exist.
+	if err := s.resolveHomeDatabase(ctx); err != nil {
+		return nil, false, err
+	}
+	serverList, err := s.getOrUpdateServers(ctx, mode, s.cache.IsEnabled())
+	return serverList, false, err
+}
+
+// borrowConnection requests a connection from the pool using the provided `serverList`.
+func (s *sessionWithContext) borrowConnection(ctx context.Context, serverList []string, livenessCheckTimeout time.Duration) (idb.Connection, error) {
+	conn, err := s.pool.Borrow(ctx, func() []string { return serverList }, s.driverConfig.ConnectionAcquisitionTimeout != 0, s.config.BoltLogger, livenessCheckTimeout, s.auth)
+	return conn, err
+}
+
+// selectDatabase ensures the correct database is selected on the connection.
+func (s *sessionWithContext) selectDatabase(ctx context.Context, conn idb.Connection) error {
+	dbSelector, ok := conn.(idb.DatabaseSelector)
+	if !ok {
+		s.pool.Return(ctx, conn)
+		return &UsageError{Message: "Database does not support multi-database"}
```

Comment:
I know this code has been here before. I think this go into a separate PR, if we decided we want to do it. While I see that code, I remember we have a specific error for such things that gets turned into a `UsageError` by `errorutil.WrapError` anyway:
```suggestion
		err := &db.FeatureNotSupportedError{
			Server: conn.ServerName(), Feature: "multi-database", Reason: "requires at least server v4",
		}
		s.pool.Return(ctx, conn)
		return err
```

Reply:
good suggestion